### PR TITLE
Refactor and merge YAML and XML loaders

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | todo

Drastically simplify the code, remove a lot of duplication and improve performances.

BC break: the class name is not the key of the definition.

Can you review this please @soyuka and @meyerbaptiste.

### CODE REVIEW

File: src/Metadata/YamlExtractor.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    /**
+     * Parses all metadata files and convert them in an array.
+     *
+     * @throws ParseException
+     *
+     * @return array
+     */
+    public function getResources(): array
```
Comment (Simperfit) [inline src/Metadata/YamlExtractor.php]: What do you think about splitting this methods ?

Comment (dunglas) [inline src/Metadata/YamlExtractor.php]: Done.

File: src/Bridge/Symfony/Bundle/Resources/config/metadata.xml [hunk 2]
```diff
@@ -41,7 +46,7 @@
         </service>
 
        <service id="api_platform.metadata.resource.metadata_factory.yaml" decorates="api_platform.metadata.resource.metadata_factory" class="ApiPlatform\Core\Metadata\Resource\Factory\YamlResourceMetadataFactory" decoration-priority="40" public="false">
-            <argument type="collection" />
+           <argument type="service" id="api_platform.metadata.extractor.yaml" />
```
Comment (meyerbaptiste) [inline src/Bridge/Symfony/Bundle/Resources/config/metadata.xml]: Bad indentation

**Comment by meyerbaptiste** [comment]
Nice, we need to do the same for XML!

**Comment by dunglas** [comment]
Ready for review. A lot of duplicated code has been removed.

**Comment by dunglas** [comment]
There is still some work to do (in other PRs):

- [ ] Test directly the extractors instead of the the factories
- [ ] Create abstract classes for factories (annotations and extractor) to avoid duplicating some methods like `handleNotFound`
- [ ] Update the YAML format in the doc (class name as key)

**Comment by soyuka** [comment]
As @teohhanhui said, I think we should keep the `class` key.

**Comment by dunglas** [comment]
Why? With the new implementation it's mandatory to have the class as key. It will just be duplicate data.

File: src/Metadata/Property/Factory/ExtractorPropertyNameCollectionFactory.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace ApiPlatform\Core\Metadata\Property\Factory;
+
+use ApiPlatform\Core\Exception\InvalidArgumentException;
+use ApiPlatform\Core\Exception\ResourceClassNotFoundException;
+use ApiPlatform\Core\Metadata\Extractor\ExtractorInterface;
+use ApiPlatform\Core\Metadata\Property\PropertyNameCollection;
+
+/**
+ * Creates a property name collection using an extractor..
```
Comment (meyerbaptiste) [inline src/Metadata/Property/Factory/ExtractorPropertyNameCollectionFactory.php]: Extra dot.

**Pull request merged**